File: bot-base-client/src/main/java/gg/mineral/bot/base/lwjgl/input/Mouse.java
package gg.mineral.bot.base.lwjgl.input;

import gg.mineral.bot.api.controls.MouseButton.Type;
import gg.mineral.bot.api.event.EventHandler;
import gg.mineral.bot.impl.config.BotGlobalConfig;

public class Mouse extends gg.mineral.bot.impl.controls.Mouse {

    public Mouse(EventHandler eventHandler) {
        super(eventHandler);
    }

    @Override
    public boolean next() {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return super.next();
        return org.lwjgl.input.Mouse.next();
    }

    @Override
    public boolean isButtonDown(int i) {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return super.next();

        return org.lwjgl.input.Mouse.isButtonDown(i);
    }

    @Override
    public boolean isButtonDown(Type type) {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return super.isButtonDown(type);

        return org.lwjgl.input.Mouse.isButtonDown(type.getKeyCode());
    }

    @Override
    public Type getEventButtonType() {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return super.getEventButtonType();

        return Type.fromKeyCode(org.lwjgl.input.Mouse.getEventButton());
    }

    @Override
    public int getEventButton() {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return super.getEventButton();

        return org.lwjgl.input.Mouse.getEventButton();
    }

    @Override
    public int getEventDWheel() {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return super.getEventDWheel();

        return org.lwjgl.input.Mouse.getEventDWheel();
    }

    @Override
    public void setX(int x) {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            super.setX(x);
        else
            org.lwjgl.input.Mouse.setCursorPosition(x, getY());
    }

    @Override
    public void setY(int y) {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            super.setY(y);
        else
            org.lwjgl.input.Mouse.setCursorPosition(getX(), y);
    }

    public void setCursorPosition(int x, int y) {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl()) {
            super.setX(x);
            super.setY(y);
        } else
            org.lwjgl.input.Mouse.setCursorPosition(x, y);
    }

    @Override
    public int getDX() {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return super.getDX();

        return org.lwjgl.input.Mouse.getDX();
    }

    @Override
    public int getDY() {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return super.getDY();

        return org.lwjgl.input.Mouse.getDY();
    }

    public boolean isCreated() {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return true;

        return org.lwjgl.input.Mouse.isCreated();
    }

    public boolean getEventButtonState() {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return super.getEventButtonState();

        return org.lwjgl.input.Mouse.getEventButtonState();
    }

    public void setGrabbed(boolean grab) {
        if (BotGlobalConfig.isHeadless() || BotGlobalConfig.isControl())
            return;

        org.lwjgl.input.Mouse.setGrabbed(grab);
    }
}

File: bot-base-client/src/main/java/gg/mineral/bot/base/lwjgl/input/Mouse.kt
package gg.mineral.bot.base.lwjgl.input

import gg.mineral.bot.api.controls.MouseButton
import gg.mineral.bot.api.controls.MouseButton.Type.Companion.fromKeyCode
import gg.mineral.bot.api.event.EventHandler
import gg.mineral.bot.impl.config.BotGlobalConfig
import gg.mineral.bot.impl.controls.Mouse

class Mouse(eventHandler: EventHandler) : Mouse(eventHandler) {
    override var isGrabbed: Boolean = false
        get() = super.isGrabbed
        set(value) {
            if (BotGlobalConfig.headless || BotGlobalConfig.control) {
                field = value
                return
            }

            org.lwjgl.input.Mouse.setGrabbed(value)
        }

    override fun next(): Boolean {
        if (BotGlobalConfig.headless || BotGlobalConfig.control) return super.next()
        return org.lwjgl.input.Mouse.next()
    }

    override fun isButtonDown(i: Int): Boolean {
        if (BotGlobalConfig.headless || BotGlobalConfig.control) return super.next()

        return org.lwjgl.input.Mouse.isButtonDown(i)
    }

    override fun isButtonDown(type: MouseButton.Type): Boolean {
        if (BotGlobalConfig.headless || BotGlobalConfig.control) return super.isButtonDown(type)

        return org.lwjgl.input.Mouse.isButtonDown(type.keyCode)
    }

    override val eventButtonType: MouseButton.Type?
        get() {
            if (BotGlobalConfig.headless || BotGlobalConfig.control) return super.eventButtonType

            return fromKeyCode(org.lwjgl.input.Mouse.getEventButton())
        }

    override val eventButton: Int
        get() {
            if (BotGlobalConfig.headless || BotGlobalConfig.control) return super.eventButton

            return org.lwjgl.input.Mouse.getEventButton()
        }

    override val eventDWheel: Int
        get() {
            if (BotGlobalConfig.headless || BotGlobalConfig.control) return super.eventDWheel

            return org.lwjgl.input.Mouse.getEventDWheel()
        }

    override var x: Int
        get() = super.x
        set(x) {
            if (BotGlobalConfig.headless || BotGlobalConfig.control) super.x =
                x
            else org.lwjgl.input.Mouse.setCursorPosition(x, y)
        }

    override var y: Int
        get() = super.y
        set(y) {
            if (BotGlobalConfig.headless || BotGlobalConfig.control) super.y =
                y
            else org.lwjgl.input.Mouse.setCursorPosition(x, y)
        }

    override fun setCursorPosition(x: Int, y: Int) {
        if (BotGlobalConfig.headless || BotGlobalConfig.control) {
            super.setCursorPosition(x, y)
        } else org.lwjgl.input.Mouse.setCursorPosition(x, y)
    }

    override var dX: Int
        get() {
            if (BotGlobalConfig.headless || BotGlobalConfig.control) return super.dX

            return org.lwjgl.input.Mouse.getDX()
        }
        set(dX) {
            super.dX = dX
        }

    override var dY: Int
        get() {
            if (BotGlobalConfig.headless || BotGlobalConfig.control) return super.dY

            return org.lwjgl.input.Mouse.getDY()
        }
        set(dY) {
            super.dY = dY
        }

    val isCreated: Boolean
        get() {
            if (BotGlobalConfig.headless || BotGlobalConfig.control) return true

            return org.lwjgl.input.Mouse.isCreated()
        }

    override val eventButtonState: Boolean
        get() {
            if (BotGlobalConfig.headless || BotGlobalConfig.control) return super.eventButtonState

            return org.lwjgl.input.Mouse.getEventButtonState()
        }
}
